design/backup-performance-improvements.md

```diff
@@ -0,0 +1,212 @@
+- Before loop iteration, create a new `itemsInBlock` map of type map[velero.ResourceIdentifier]bool which represents the set of items already included in a block.
+- If `item` is already in `itemsInBlock`, continue. This one has already been processed.
+- Add `item` to `itemsInBlock`.
+- Load item from ItemCollector file. Close/remove file after loading (on error return or not, possibly with similar anonymous func to current impl)
+- Get matching BIA plugins for item, call `GetAdditionalItems` for each. For each item returned, get full item content from ItemCollector (if present in item list, pulling from file, removing file when done) or from cluster (if not present in item list), add item to the current block, add item to `itemsInBlock` map, and then recursively apply current step to each (i.e. call BIA method, add to block, etc.)
+- Once full ItemBlock list is generated, call `backupItemBlock(block ItemBlock)
+- Add `backupItemBlock` return values to `backedUpGroupResources` map
+
+Note that if there are BIAv2 plugins present, we will downgrade all v3 plugins to v2, meaning that we will not have a `GetAdditionalItems` func to call. In this case, we will make the following change to the ItemCollector iteration workflow described above:
+- At the point where we call `GetAdditionalItems` on the item's registered plugins, we will not call this. Instead, we will treat the item as if it had returned no additional items. The ItemBlock will only have one entry.
+
+#### New func `backupItemBlock`
+
+Method signature for new func `backupItemBlock` is as follows:
+```
+func backupItemBlock(block ItemBlock) []schema.GroupResource
+```
+The return value is a slice of GRs for resources which were backed up. Velero tracks these to determine which CRDs need to be included in the backup. Note that we need to make sure we include in this not only those resources that were backed up directly, but also those backed up indirectly via additional items BIA execute returns.
+
+In order to handle backup hooks, this func will first take the input item list (`block.items`) and get a list of included pods, filtered to include only those not yet backed up (using `block.itemBackupper.backupRequest.BackedUpItems`). Iterate over this list and execute pre hooks (pulled out of `itemBackupper.backupItemInternal`) for each item.
+Now iterate over the full list (`block.items`) and call `backupItem` for each. After the first, the later items should already have been backed up, but calling a second time is harmless, since the first thing Velero does is check the `BackedUpItems` map, exiting if item is already backed up). We still need this call in case there's a plugin which returns something in `GetAdditionalItems` but forgets to return it in the `Execute` additional items return value. If we don't do this, we could end up missing items.
```

Comment:
This allows us to preserve the logic within BackupItem, where the understanding is that items returned here must be backed up *right now*. We could relax this and say "if you have custom plugin logic that depends on the order of processing within the block, you need to return additional items in both GetAdditionalItems and Execute."

Reply:
Thinking through a concrete example. The most complicated relationship now is with pods. Current logic:
```
- pod backup starts, pre hooks run
- pod plugins called, returns PVC list as additional items
- start pvc backup
  - pvc plugins called, creates VS and returns VS as additional item
  - start vs backup, creates VSC and returns VSC as additional item
    - start VSC backup
      - VSC plugins called
    - finish VSC backup
  - finish VS backup
- finish PVC backup
- finish pod backup
- call pod post hooks  
```

Reply:
New logic if we don't require plugins to return additional items during execute if they already returned them during GetAdditionalItems:
```
- Create item block for pod:
  - Call GetAdditionalItems for pod plugins, returns PVC
    - Call GetAdditionalItems for PVC plugins, returns PV (but not VS, VSC, etc)
- BackupItemBlock for block containing pod, PVC, PV
- pre hooks run
- start pod backup
- pod plugins called, no additional items
- end pod backup
- start PVC backup
- pvc plugins called, creates VS and returns VS as additional item
  - start vs backup, creates VSC and returns VSC as additional item
    - start VSC backup
      - VSC plugins called
    - finish vsc backup
- finish PVC backup
- start PV backup, finish PV backup
- call post hooks
```

Reply:
On the surface, it looks like the modified approach (not requiring Execute() returning items already returned in preprocessing) should work, but with one important caveat. If we don't require Execute() to return *all* additional items, then when we adapt a v3 BIA to v2 (in cases where we must fall back to prior behavior), the adapter for Execute() must call both GetAdditionalItems and Execute and merge the item lists into the Execute output, otherwise we'll miss backing up the PVC and PV at the right time.

Reply:
@shawn-hurley Any thoughts on the above?

Reply:
The major reason is that v3 plugins are backward compatible; changing this would break this compatibility and would be undesirable, in my opinion.

Reply:
@shawn-hurley It wouldn't break the compatibility, since the adapter would handle the required downgrade to v2 in cases where we have v2 plugins registered -- the adapter would make sure that Execute() returns all additional items in the case where it's being called in a v2 downgrade environment. While it makes the adapter slightly more complicated, it does make it easier to prevent certain plugin bugs -- with the current design, if a plugin writer includes items in GetAdditionalItems but not in execute, then it will work properly in the v3 use case, but it will fail to work properly in the v2 downgrade case because items will be missed.